File: leetcode/109-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ret = []
        if root is None:
            return ret

        q = [root]
        while len(q) > 0:
            ret += [q[-1].val]
            q = [child
                 for node in q
                 for child in [node.left, node.right]
                 if child is not None]
        return ret

File: leetcode/109-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  vector<int> rightSideView(TreeNode* root) {
    vector<int> ret;
    if (!root) return ret;

    queue<TreeNode*> q;
    TreeNode* sentinal = new TreeNode(0);
    int val = 0;
    q.push(root);
    q.push(sentinal);
    for (TreeNode* n; !q.empty(); ) {
      n = q.front(); q.pop();
      if (n == sentinal) {
        ret.push_back(val);
        if (q.empty()) break;
        q.push(sentinal);
      } else {
        val = n->val;
        if (n->left) q.push(n->left);
        if (n->right) q.push(n->right);
      }
    }
    return ret;
  }
};
